src/audio/mfcc/mfcc_generic.c

```diff
@@ -0,0 +1,496 @@
+	int n1;
+	int n2;
+	int n;
+	int i;
+	int num_channels = source->channels;
+
+	/* Copy from source to pre-buffer for FFT.
+	 * The pre-emphasis filter is done in this step.
+	 */
+	for (copied = 0; copied < frames; copied += n) {
+		nmax = frames - copied;
+		n1 = audio_stream_frames_without_wrap(source, x);
+		n2 = mfcc_buffer_samples_without_wrap(buf, w);
+		n = MIN(n1, n2);
+		n = MIN(n, nmax);
+		x0 = x + source_channel;
+		for (i = 0; i < n; i++) {
+			if (emph->enable) {
+				/* Q1.15 x Q1.15 -> Q2.30 */
+				s = (int32_t)emph->delay * emph->coef + Q_SHIFT_LEFT(*x0, 15, 30);
+				*w = sat_int16(Q_SHIFT_RND(s, 30, 15));
+				emph->delay = *x0;
+			} else {
+				*w = *x0;
+			}
+			x0 += num_channels;
+			w++;
+		}
+
+		x = audio_stream_wrap(source, x + n * source->channels);
+		w = mfcc_buffer_wrap(buf, w);
+	}
+	buf->s_avail += copied;
+	buf->s_free -= copied;
+	buf->w_ptr = w;
+}
+
+static void mfcc_fill_prev_samples(struct mfcc_buffer *buf, int16_t *prev_data,
+				   int prev_data_length)
+{
+	/* Fill prev_data from input buffer */
+	int16_t *r = buf->r_ptr;
+	int16_t *p = prev_data;
+	int copied;
+	int nmax;
+	int n;
+
+	for (copied = 0; copied < prev_data_length; copied += n) {
+		nmax = prev_data_length - copied;
+		n = mfcc_buffer_samples_without_wrap(buf, r);
+		n = MIN(n, nmax);
+		memcpy(p, r, sizeof(int16_t) * n); /* Not using memcpy_s() due to speed need */
+		p += n;
+		r += n;
+		r = mfcc_buffer_wrap(buf, r);
+	}
+
+	buf->s_avail -= copied;
+	buf->s_free += copied;
+	buf->r_ptr = r;
+}
+
+static void mfcc_fill_fft_buffer(struct mfcc_state *state)
+{
+	struct mfcc_buffer *buf = &state->buf;
+	struct mfcc_fft *fft = &state->fft;
+	int16_t *r = buf->r_ptr;
+	int copied;
+	int nmax;
+	int idx = fft->fft_fill_start_idx;
+	int j;
+	int n;
+
+	/* Copy overlapped samples from state buffer. Imaginary part of input
+	 * remains zero.
+	 */
+	for (j = 0; j < state->prev_data_size; j++)
+		fft->fft_buf[idx + j].real = state->prev_data[j];
+
+	/* Copy hop size of new data from circular buffer */
+	idx += state->prev_data_size;
+	for (copied = 0; copied < fft->fft_hop_size; copied += n) {
+		nmax = fft->fft_hop_size - copied;
+		n = mfcc_buffer_samples_without_wrap(buf, r);
+		n = MIN(n, nmax);
+		for (j = 0; j < n; j++) {
+			fft->fft_buf[idx].real = *r;
+			r++;
+			idx++;
+		}
+		r = mfcc_buffer_wrap(buf, r);
+	}
+
+	buf->s_avail -= copied;
+	buf->s_free += copied;
+	buf->r_ptr = r;
+
+	/* Copy for next time data back to overlap buffer */
+	idx = fft->fft_fill_start_idx + fft->fft_hop_size;
+	for (j = 0; j < state->prev_data_size; j++)
+		state->prev_data[j] = fft->fft_buf[idx + j].real;
+}
+
+#ifdef MFCC_NORMALIZE_FFT
+static int mfcc_normalize_fft_buffer(struct mfcc_state *state)
+{
+	struct mfcc_fft *fft = &state->fft;
+	int32_t absx;
+	int32_t smax = 0;
+	int32_t x;
+	int shift;
+	int j;
+	int i = fft->fft_fill_start_idx;
+
+	for (j = 0; j < fft->fft_size; j++) {
+		x = fft->fft_buf[i + j].real;
+		absx = (x < 0) ? -x : x;
```

Comment:
why not use `ABS()` from src/include/sof/math/numbers.h ?

Reply:
This is hot path, the macro is more than above line, is there overhead from typeof() typecast operation there? Also I wonder if Zephyr comes with a different ABS()?

Reply:
the potential overhead from using `ABS()`  would be not from using `typeof()` but from allocating a new variable on stack. `ABS()` does the right thing with that to avoid repeated expression evaluation. E.g. it's safe to do `ABS(f(x))` whereas if `ABS()` were just a trivial `#define ABS(x) ((x) < 0 ? -(x) : (x))` that would lead to evaluating `f(x)` twice which can have undesirable side effects. Whereas the SOF `ABS(x)` implementation avoids that. In the above case it isn't needed since `x` is a simple variable, but it's important for a generic macro. I would guess that the compiler would optimise that additional variable on stack out, but if you're concerned about that - I'd understand that too.

Reply:
Ah for that, makes sense. In the beginning we made for most things macros but I've been wondering if inline functions would be better to avoid unknowns. Inlines would be also easier for seamless intrinsic replace. I'd like to keep this as current. Also mfcc_generic.c would be later converted with intrinsics.